pkg/sql/schemachanger/scplan/internal/rules/current/dep_swap_index.go

```diff
@@ -125,3 +124,60 @@ func init() {
 		},
 	)
 }
+
+// This rule ensures that when secondary indexes are re-created after a primary
+// index key is changed, that the secondary indexes are swapped in an atomic
+// manner, so that queries are not impacted by missing indexes.
+func init() {
+	// This ia strict version of the rule that will only work, when a node
+	// is generating a plan on the latest master / 23.1. The StrictRecreate flag
+	// will be used to tag if the existing secondary index was created on a
+	// new enough version.
+	registerDepRule(
+		"secondary index being recreated after a primary key swap",
+		scgraph.SameStagePrecedence,
```

Comment:
I see why you use SameStagePrecedence here but do we need to? I just have some invalidated fear from past experience so that I try to avoid SameStagePrecedence if possible.

What about a Precedence from (new_index, public) to (old_index, validated)? This is saying to the old index "hey, sorry, you cannot transition to VALIDATED, until your new index is PUBLIC". The planner, being a greedy one, should naturally schedule both transitions to the same stage when new_index transitions to PUBLIC.